include/oneapi/mkl/detail/backend_selector.hpp

```diff
@@ -0,0 +1,80 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions
+* and limitations under the License.
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+#ifndef _ONEMKL_BACKEND_SELECTOR_HPP_
+#define _ONEMKL_BACKEND_SELECTOR_HPP_
+
+#include "oneapi/mkl/exceptions.hpp"
+#include "oneapi/mkl/detail/backends.hpp"
+#include "oneapi/mkl/detail/get_device_id.hpp"
+
+namespace oneapi {
+namespace mkl {
+
+template <backend Backend>
+class backend_selector {
+public:
+    explicit backend_selector(sycl::queue queue) : queue_(queue) {
+        if ((queue.is_host() || queue.get_device().is_cpu())) {
```

Comment:
Hi @aelizaro -- do you think it makes sense to have a `switch` on the backend type, then have an (inline) function for each backend (defined in the backend-specific code) to check whether it can handle the given device? This can save on a few SYCL API calls and also make things a bit more modular as each backend could "own" its device check code more easily.

Reply:
Hi @petercad, thank you for the idea! I have reworked this part: a new file with checks baskend_selector_predicates.hpp is added. I made inlined functions with the backend template parameter, and it may be overloaded for each backend and it completes check only if ONEMKL_DISABLE_PREDICATES is not specified.

Reply:
Looks nice -- thanks @aelizaro!